pkg/registry/generic/rest/proxy.go

```diff
@@ -76,12 +76,33 @@ func (h *UpgradeAwareProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Re
 		return
 	}
 
+	loc := *h.Location
+	loc.RawQuery = req.URL.RawQuery
+
+	// If original request URL ended in '/', append a '/' at the end of the
+	// of the proxy URL
+	if !strings.HasSuffix(loc.Path, "/") && strings.HasSuffix(req.URL.Path, "/") {
+		loc.Path += "/"
+	}
+
+	// From pkg/apiserver/proxy.go#ServeHTTP:
+	// Redirect requests with an empty path to a location that ends with a '/'
+	// This is essentially a hack for https://github.com/GoogleCloudPlatform/kubernetes/issues/4958.
+	// Note: Keep this code after tryUpgrade to not break that flow.
+	if len(loc.Path) == 0 {
+		var queryPart string
+		if len(req.URL.RawQuery) > 0 {
+			queryPart = "?" + req.URL.RawQuery
+		}
+		w.Header().Set("Location", req.URL.Path+"/"+queryPart)
```

Comment:
we're not including a host in the Location... and from what I can tell, that is a valid relative URL (http://en.wikipedia.org/wiki/HTTP_location)
It's also what we use in the original proxy code: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/apiserver/proxy.go#L166

You'd also be relying on the reverse proxy or whatever's in front of your server to translate the header properly and not expose the name of your internal server. So for example your might hit a load balancer at http://foo/api/v1beta3/namespaces/default/services/foo/proxy, but the URL api server sees is http://bar/api/v1beta3/namespaces/default/services/foo/proxy, so we redirect to http://bar/api/v1beta3/namespaces/default/services/foo/proxy/ and if it comes back unfiltered to the client, the client may not know how to access http://bar/...